lib/lambda_ethereum_consensus/validator/validator_set.ex

```diff
@@ -0,0 +1,287 @@
+    Logger.warning(
+      "[Validator] No keystore_dir or keystore_pass_dir provided. Validators won't start."
+    )
+
+    %__MODULE__{}
+  end
+
+  defp setup_validators(slot, head_root, keystore_dir, keystore_pass_dir) do
+    validator_keystores = decode_validator_keystores(keystore_dir, keystore_pass_dir)
+    epoch = Misc.compute_epoch_at_slot(slot)
+
+    validators =
+      Map.new(validator_keystores, fn keystore ->
+        validator = Validator.new(keystore, slot, head_root)
+        {validator.index, validator}
+      end)
+
+    Logger.info("[Validator] Initialized #{Enum.count(validators)} validators")
+
+    %__MODULE__{validators: validators}
+    |> update_state(epoch, slot, head_root)
+  end
+
+  @doc """
+  Notify all validators of a new head.
+  """
+  @spec notify_head(t(), Types.slot(), Types.root()) :: t()
+  def notify_head(set, slot, head_root) do
+    # TODO: Just for testing purposes, remove it later
+    Logger.info("[ValidatorSet] New Head", root: head_root, slot: slot)
+    epoch = Misc.compute_epoch_at_slot(slot)
+
+    set
+    |> update_state(epoch, slot, head_root)
+    |> attests(epoch, slot, head_root)
+    |> build_payload(slot + 1, head_root)
+  end
+
+  @doc """
+  Notify all validators of a new tick.
+  """
+  @spec notify_tick(t(), tuple()) :: t()
+  def notify_tick(%{head_root: head_root} = set, {slot, third} = slot_data) do
+    # TODO: Just for testing purposes, remove it later
+    Logger.info("[ValidatorSet] Tick #{inspect(third)}", root: head_root, slot: slot)
+    epoch = Misc.compute_epoch_at_slot(slot)
+
+    set
+    |> update_state(epoch, slot, head_root)
+    |> process_tick(epoch, slot_data)
+  end
+
+  defp process_tick(%{head_root: head_root} = set, epoch, {slot, :first_third}) do
+    propose(set, epoch, slot, head_root)
+  end
+
+  defp process_tick(%{head_root: head_root} = set, epoch, {slot, :second_third}) do
+    set
+    |> attests(epoch, slot, head_root)
+    |> build_payload(slot + 1, head_root)
+  end
+
+  defp process_tick(set, epoch, {slot, :last_third}) do
+    publish_aggregates(set, epoch, slot)
+  end
+
+  ##############################
+  # State update
+
+  defp update_state(set, epoch, slot, head_root) do
+    set
+    |> update_head(head_root)
+    |> compute_duties(epoch, slot, head_root)
+  end
+
+  defp update_head(%{head_root: head_root} = set, head_root), do: set
+  defp update_head(set, head_root), do: %{set | head_root: head_root}
+
+  defp compute_duties(set, epoch, _slot, _head_root)
+       when is_duties_computed(set, epoch) and is_duties_computed(set, epoch + 1),
+       do: set
+
+  defp compute_duties(set, epoch, slot, head_root) do
+    epochs_to_calculate =
+      [{epoch, slot}, {epoch + 1, Misc.compute_start_slot_at_epoch(epoch + 1)}]
+      |> Enum.reject(&Map.has_key?(set.duties, elem(&1, 0)))
+
+    epochs_to_calculate
+    |> Map.new(&compute_duties_for_epoch!(set, &1, head_root))
+    |> merge_duties_and_prune(epoch, set)
```

Comment:
This is assuming (as it was done previously) that we can calculate duties ahead of time, this shouldn't be the case because Validators participating in an epoch will be decided at the start of that epoch.

Reply:
From the randao chapter in the [eth2book](https://eth2book.info/capella/part2/building_blocks/randomness/#lookahead)

>The RANDAO seed at the end of epoch  N is used to compute validator duties for the whole of epoch N+2. This interval is controlled by [MIN_SEED_LOOKAHEAD](https://eth2book.info/capella/part3/config/preset/#min_seed_lookahead) via the [get_seed()](https://eth2book.info/capella/part3/helper/accessors/#def_get_seed) function. Thus, validators have at least one full epoch to prepare themselves for any duties, but no more than two.

Reply:
Thanks Tomy! I was wondering about validator's activations and exits when thinking about this, like wouldn't new activations or exits affect already calculated duties with a different set of validators? I might be lacking more knowledge about RANDAO and how the lookahead plays with the calculation of duties for te active set of validators.